File: composeApp/src/androidMain/kotlin/com/kissangram/repository/MockPostRepository.kt
package com.kissangram.repository

import com.kissangram.model.*
import com.kissangram.repository.PostRepository
import kotlinx.coroutines.delay

/**
 * Mock implementation of PostRepository with dummy data for development
 */
class MockPostRepository : PostRepository {
    
    // In-memory state for likes and saves (would be persisted in real implementation)
    private val likedPosts = mutableSetOf<String>()
    private val savedPosts = mutableSetOf<String>()
    
    override suspend fun getPost(postId: String): Post? {
        delay(300)
        // Return a sample post
        return null
    }
    
    override suspend fun likePost(postId: String) {
        delay(200)
        likedPosts.add(postId)
    }
    
    override suspend fun unlikePost(postId: String) {
        delay(200)
        likedPosts.remove(postId)
    }
    
    override suspend fun savePost(postId: String) {
        delay(200)
        savedPosts.add(postId)
    }
    
    override suspend fun unsavePost(postId: String) {
        delay(200)
        savedPosts.remove(postId)
    }
    
    override suspend fun getComments(postId: String, page: Int, pageSize: Int): List<Comment> {
        delay(400)
        
        val currentTime = System.currentTimeMillis()
        
        return listOf(
            Comment(
                id = "comment1",
                postId = postId,
                authorId = "user10",
                authorName = "Dr. Sharma",
                authorUsername = "dr_sharma",
                authorProfileImageUrl = "https://images.unsplash.com/photo-1612349317150-e413f6a5b16d?w=150",
                authorRole = UserRole.EXPERT,
                authorVerificationStatus = VerificationStatus.VERIFIED,
                text = "Great progress! Make sure to check soil moisture before next irrigation.",
                voiceComment = null,
                parentCommentId = null,
                repliesCount = 2,
                likesCount = 15,
                isLikedByMe = false,
                isExpertAnswer = true,
                isBestAnswer = false,
                createdAt = currentTime - 1800000
            ),
            Comment(
                id = "comment2",
                postId = postId,
                authorId = "user11",
                authorName = "Anil Kumar",
                authorUsername = "anil_farmer",
                authorProfileImageUrl = "https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?w=150",
                authorRole = UserRole.FARMER,
                authorVerificationStatus = VerificationStatus.UNVERIFIED,
                text = "Which variety is this? I want to try next season.",
                voiceComment = null,
                parentCommentId = null,
                repliesCount = 1,
                likesCount = 5,
                isLikedByMe = true,
                isExpertAnswer = false,
                isBestAnswer = false,
                createdAt = currentTime - 3600000
            ),
            Comment(
                id = "comment3",
                postId = postId,
                authorId = "user12",
                authorName = "Ravi Singh",
                authorUsername = "ravi_agri",
                authorProfileImageUrl = null,
                authorRole = UserRole.FARMER,
                authorVerificationStatus = VerificationStatus.UNVERIFIED,
                text = "Very nice! What's your irrigation schedule?",
                voiceComment = null,
                parentCommentId = null,
                repliesCount = 0,
                likesCount = 3,
                isLikedByMe = false,
                isExpertAnswer = false,
                isBestAnswer = false,
                createdAt = currentTime - 5400000
            )
        )
    }
    
    override suspend fun addComment(postId: String, text: String): Comment {
        delay(500)
        
        return Comment(
            id = "comment_new_${System.currentTimeMillis()}",
            postId = postId,
            authorId = "current_user",
            authorName = "You",
            authorUsername = "current_user",
            authorProfileImageUrl = null,
            authorRole = UserRole.FARMER,
            authorVerificationStatus = VerificationStatus.UNVERIFIED,
            text = text,
            voiceComment = null,
            parentCommentId = null,
            repliesCount = 0,
            likesCount = 0,
            isLikedByMe = false,
            isExpertAnswer = false,
            isBestAnswer = false,
            createdAt = System.currentTimeMillis()
        )
    }
    
    override suspend fun getPostsByUser(userId: String, page: Int, pageSize: Int): List<Post> {
        delay(400)
        return emptyList()
    }
    
    override suspend fun getPostsByCrop(crop: String, page: Int, pageSize: Int): List<Post> {
        delay(400)
        return emptyList()
    }
    
    override suspend fun getPostsByHashtag(hashtag: String, page: Int, pageSize: Int): List<Post> {
        delay(400)
        return emptyList()
    }
    
    override suspend fun createPost(postData: Map<String, Any?>): Post {
        delay(500) // Simulate network delay
        
        val postId = "mock_post_${System.currentTimeMillis()}"
        val authorId = postData["authorId"] as? String ?: "mock_user"
        val authorName = postData["authorName"] as? String ?: "Mock User"
        val authorUsername = postData["authorUsername"] as? String ?: "mock_user"
        val authorProfileImageUrl = postData["authorProfileImageUrl"] as? String
        val text = postData["text"] as? String ?: ""
        
        // Parse media array
        val mediaList = (postData["media"] as? List<Map<String, Any>>)?.mapNotNull { mediaMap ->
            val url = mediaMap["url"] as? String ?: return@mapNotNull null
            val typeStr = mediaMap["type"] as? String ?: "image"
            val thumbnailUrl = mediaMap["thumbnailUrl"] as? String
            
            PostMedia(
                url = url,
                type = when (typeStr) {
                    "video" -> MediaType.VIDEO
                    else -> MediaType.IMAGE
                },
                thumbnailUrl = thumbnailUrl?.takeIf { it.isNotEmpty() }
            )
        } ?: emptyList()
        
        // Parse voice caption
        val voiceCaptionMap = postData["voiceCaption"] as? Map<String, Any>
        val voiceCaption = voiceCaptionMap?.let {
            VoiceContent(
                url = it["url"] as? String ?: return@let null,
                durationSeconds = (it["durationSeconds"] as? Number)?.toInt() ?: 0
            )
        }
        
        // Parse crops and hashtags
        val crops = (postData["crops"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
        val hashtags = (postData["hashtags"] as? List<*>)?.mapNotNull { it as? String } ?: emptyList()
        
        // Parse location
        val locationMap = postData["location"] as? Map<String, Any>
        val location = locationMap?.let {
            PostLocation(
                name = it["name"] as? String ?: "",
                latitude = it["latitude"] as? Double,
                longitude = it["longitude"] as? Double
            )
        }
        
        // Parse question data
        val questionMap = postData["question"] as? Map<String, Any>
        val question = questionMap?.let {
            QuestionData(
                targetExpertise = (it["targetExpertise"] as? List<*>)?.mapNotNull { e -> e as? String } ?: emptyList(),
                targetExpertIds = (it["targetExpertIds"] as? List<*>)?.mapNotNull { e -> e as? String } ?: emptyList(),
                targetExperts = emptyList(), // Mock doesn't populate this
                isAnswered = it["isAnswered"] as? Boolean ?: false,
                bestAnswerCommentId = it["bestAnswerCommentId"] as? String
            )
        }
        
        val typeStr = postData["type"] as? String ?: "normal"
        val roleStr = postData["authorRole"] as? String ?: "farmer"
        val verificationStatusStr = postData["authorVerificationStatus"] as? String ?: "unverified"
        
        return Post(
            id = postId,
            authorId = authorId,
            authorName = authorName,
            authorUsername = authorUsername,
            authorProfileImageUrl = authorProfileImageUrl,
            authorRole = when (roleStr) {
                "expert" -> UserRole.EXPERT
                "agripreneur" -> UserRole.AGRIPRENEUR
                "input_seller" -> UserRole.INPUT_SELLER
                "agri_lover" -> UserRole.AGRI_LOVER
                else -> UserRole.FARMER
            },
            authorVerificationStatus = when (verificationStatusStr) {
                "pending" -> VerificationStatus.PENDING
                "verified" -> VerificationStatus.VERIFIED
                "rejected" -> VerificationStatus.REJECTED
                else -> VerificationStatus.UNVERIFIED
            },
            type = when (typeStr) {
                "question" -> PostType.QUESTION
                else -> PostType.NORMAL
            },
            text = text,
            media = mediaList,
            voiceCaption = voiceCaption,
            crops = crops,
            hashtags = hashtags,
            location = location,
            question = question,
            likesCount = 0,
            commentsCount = 0,
            savesCount = 0,
            isLikedByMe = false,
            isSavedByMe = false,
            createdAt = System.currentTimeMillis(),
            updatedAt = null
        )
    }
}

File: iosApp/iosApp/repository/MockPostRepository.swift
import Foundation
import Shared

/**
 * Mock implementation of PostRepository with dummy data for development
 */
final class MockPostRepository: PostRepository {
    
    private var likedPosts = Set<String>()
    private var savedPosts = Set<String>()
    
    func getPost(postId: String) async throws -> Post? {
        try? await Task.sleep(nanoseconds: 300_000_000)
        return nil
    }
    
    func likePost(postId: String) async throws {
        try? await Task.sleep(nanoseconds: 200_000_000)
        likedPosts.insert(postId)
    }
    
    func unlikePost(postId: String) async throws {
        try? await Task.sleep(nanoseconds: 200_000_000)
        likedPosts.remove(postId)
    }
    
    func savePost(postId: String) async throws {
        try? await Task.sleep(nanoseconds: 200_000_000)
        savedPosts.insert(postId)
    }
    
    func unsavePost(postId: String) async throws {
        try? await Task.sleep(nanoseconds: 200_000_000)
        savedPosts.remove(postId)
    }
    
    func getComments(postId: String, page: Int32, pageSize: Int32) async throws -> [Comment] {
        try? await Task.sleep(nanoseconds: 400_000_000)
        // Return empty list - use FirestorePostRepository for actual data
        return []
    }
    
    func addComment(postId: String, text: String, parentCommentId: String?) async throws -> Comment {
        try? await Task.sleep(nanoseconds: 500_000_000)
        
        return Comment(
            id: "comment_new_\(Int(Date().timeIntervalSince1970 * 1000))",
            postId: postId,
            authorId: "current_user",
            authorName: "You",
            authorUsername: "current_user",
            authorProfileImageUrl: nil,
            authorRole: .farmer,
            authorVerificationStatus: .unverified,
            text: text,
            voiceComment: nil,
            parentCommentId: parentCommentId,
            repliesCount: Int32(0),
            likesCount: Int32(0),
            isLikedByMe: false,
            isExpertAnswer: false,
            isBestAnswer: false,
            createdAt: Int64(Date().timeIntervalSince1970 * 1000)
        )
    }
    
    func deleteComment(postId: String, commentId: String, reason: String) async throws {
        try? await Task.sleep(nanoseconds: 300_000_000)
        // Mock implementation - just simulate deletion
    }
    
    func getPostsByUser(userId: String, page: Int32, pageSize: Int32) async throws -> [Post] {
        try? await Task.sleep(nanoseconds: 400_000_000)
        return []
    }
    
    func getPostsByCrop(crop: String, page: Int32, pageSize: Int32) async throws -> [Post] {
        try? await Task.sleep(nanoseconds: 400_000_000)
        return []
    }
    
    func getPostsByHashtag(hashtag: String, page: Int32, pageSize: Int32) async throws -> [Post] {
        try? await Task.sleep(nanoseconds: 400_000_000)
        return []
    }
    
    func createPost(postData: [String: Any], completionHandler: @escaping (Post?, Error?) -> Void) {
        // Convert [String: Any] to [String: Any?] for internal use
        let postDataWithOptionals: [String: Any?] = postData.mapValues { $0 }
        
        Task {
            do {
                try? await Task.sleep(nanoseconds: 500_000_000) // Simulate network delay
                
                let postId = "mock_post_\(Int(Date().timeIntervalSince1970 * 1000))"
                let authorId = postDataWithOptionals["authorId"] as? String ?? "mock_user"
                let authorName = postDataWithOptionals["authorName"] as? String ?? "Mock User"
                let authorUsername = postDataWithOptionals["authorUsername"] as? String ?? "mock_user"
                let authorProfileImageUrl = postDataWithOptionals["authorProfileImageUrl"] as? String
                let text = postDataWithOptionals["text"] as? String ?? ""
                
                // Parse media array
                let mediaArray = (postDataWithOptionals["media"] as? [[String: Any]])?.compactMap { mediaMap -> PostMedia? in
                    guard let url = mediaMap["url"] as? String else { return nil }
                    let typeStr = mediaMap["type"] as? String ?? "image"
                    let thumbnailUrl = mediaMap["thumbnailUrl"] as? String
                    
                    return PostMedia(
                        url: url,
                        type: typeStr == "video" ? .video : .image,
                        thumbnailUrl: thumbnailUrl?.isEmpty == false ? thumbnailUrl : nil
                    )
                } ?? []
                
                // Parse voice caption
                let voiceCaptionMap = postDataWithOptionals["voiceCaption"] as? [String: Any]
                let voiceCaption: VoiceContent? = voiceCaptionMap.flatMap { map in
                    guard let url = map["url"] as? String else { return nil }
                    let duration = (map["durationSeconds"] as? Int) ?? 0
                    return VoiceContent(url: url, durationSeconds: Int32(duration))
                }
                
                // Parse crops and hashtags
                let crops = (postDataWithOptionals["crops"] as? [String]) ?? []
                let hashtags = (postDataWithOptionals["hashtags"] as? [String]) ?? []
                
                // Parse location
                let locationMap = postDataWithOptionals["location"] as? [String: Any]
                let location: PostLocation? = locationMap.flatMap { map in
                    let name = map["name"] as? String ?? ""
                    let latitude = map["latitude"] as? Double
                    let longitude = map["longitude"] as? Double
                    return PostLocation(
                        name: name,
                        latitude: latitude != nil ? KotlinDouble(value: latitude!) : nil,
                        longitude: longitude != nil ? KotlinDouble(value: longitude!) : nil
                    )
                }
                
                // Parse question data
                let questionMap = postDataWithOptionals["question"] as? [String: Any]
                let question: QuestionData? = questionMap.flatMap { map in
                    let targetExpertise = (map["targetExpertise"] as? [String]) ?? []
                    let targetExpertIds = (map["targetExpertIds"] as? [String]) ?? []
                    return QuestionData(
                        targetExpertise: targetExpertise,
                        targetExpertIds: targetExpertIds,
                        targetExperts: [],
                        isAnswered: (map["isAnswered"] as? Bool) ?? false,
                        bestAnswerCommentId: map["bestAnswerCommentId"] as? String
                    )
                }
                
                let typeStr = postDataWithOptionals["type"] as? String ?? "normal"
                let roleStr = postDataWithOptionals["authorRole"] as? String ?? "farmer"
                let verificationStatusStr = postDataWithOptionals["authorVerificationStatus"] as? String ?? "unverified"
                
                let post = Post(
                    id: postId,
                    authorId: authorId,
                    authorName: authorName,
                    authorUsername: authorUsername,
                    authorProfileImageUrl: authorProfileImageUrl,
                    authorRole: stringToUserRole(roleStr),
                    authorVerificationStatus: stringToVerificationStatus(verificationStatusStr),
                    type: typeStr == "question" ? .question : .normal,
                    text: text,
                    media: mediaArray,
                    voiceCaption: voiceCaption,
                    crops: crops,
                    hashtags: hashtags,
                    location: location,
                    question: question,
                    likesCount: 0,
                    commentsCount: 0,
                    savesCount: 0,
                    isLikedByMe: false,
                    isSavedByMe: false,
                    createdAt: Int64(Date().timeIntervalSince1970 * 1000),
                    updatedAt: nil
                )
                
                DispatchQueue.main.async {
                    completionHandler(post, nil)
                }
            } catch {
                DispatchQueue.main.async {
                    completionHandler(nil, error)
                }
            }
        }
    }
    
    private func stringToUserRole(_ roleStr: String) -> UserRole {
        switch roleStr {
        case "expert": return .expert
        case "agripreneur": return .agripreneur
        case "input_seller": return .inputSeller
        case "agri_lover": return .agriLover
        default: return .farmer
        }
    }
    
    private func stringToVerificationStatus(_ statusStr: String) -> VerificationStatus {
        switch statusStr {
        case "pending": return .pending
        case "verified": return .verified
        case "rejected": return .rejected
        default: return .unverified
        }
    }
}
